File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include<iostream>
#include<stack>
#include<string>
#pragma warning (disable:4996)

using namespace std;

int main() {

	while (1) {
		string input;
		stack<char> s;
		int check = 0;
		getline(cin, input);
		if (input == ".") {
			break;
		}

		for (int i = 0; i < input.size(); i++) {
			if (input[i] == '(' || input[i] == '[') {
				s.push(input[i]);
			}
			else if (input[i]==')') {
				if (!s.empty() && s.top() == '(') {
					s.pop();
				}
				else {
					check = 1;
					break;
				}
			}
			else if (input[i] == ']') {
				if (!s.empty() && s.top() == '[') {
					s.pop();
				}
				else {
					check = 1;
					break;
				}
			}
		}
		if (check == 0 && s.empty()) {
			printf("yes\n");
		}
		else {
			printf("no\n");
		}
	}
	
	return 0;
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java

import java.io.*;
import java.util.*;


public class Main {
    int n,answer;
    String input;
    boolean flag;

    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;

    Stack<Character> stk = new Stack<>();

    void solution() throws Exception{
        while(true){
            input=br.readLine();
            flag=true;
            int len=input.length();

            if(len==1&&input.charAt(0)=='.'){
                break;
            }

            for(int i=0;i<len;i++){
                char c=input.charAt(i);
                if(c=='('||c=='[')
                    stk.push(c);
                else if(c==')'){
                    if(!stk.isEmpty()&&stk.peek()=='(') {
                        stk.pop();
                    }
                    else{
                        flag=false;
                        break;
                    }
                }
                else if(c==']'){
                    if(!stk.isEmpty()&&stk.peek()=='[') {
                        stk.pop();
                    }
                    else{
                        flag=false;
                        break;
                    }
                }
            }
            if(stk.isEmpty()&&flag)
                System.out.println("yes");
            else
                System.out.println("no");
            stk.clear();
        }
    }



    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}